### PR TITLE
Moved client cert checking into Authenticator

### PR DESCRIPTION
**Pull request by chirino**
Also support enabling client cert and other auth modes at the same time.

### CODE REVIEW

File: agent/jvm/src/main/java/org/jolokia/jvmagent/security/ClientCertAuthenticator.java [hunk 1]
```diff
@@ -112,16 +116,13 @@ private void checkCertForAllowedPrincipals(HttpsExchange pHttpsExchange) {
...
                 Set<Rdn> certPrincipalRdns = getPrincipalRdns(certPrincipal);
-                boolean matchFound = false;
                 for (LdapName principal : allowedPrincipals) {
-                    if( certPrincipalRdns.containsAll(principal.getRdns()) ) {
-                        matchFound = true;
-                        break;
+                    for (Rdn rdn : principal.getRdns()) {
+                        if (!certPrincipalRdns.contains(rdn)) {
+                            throw new SecurityException("Principal " + certPrincipal + " not allowed");
+                        }
```
Comment (rhuss) [inline agent/jvm/src/main/java/org/jolokia/jvmagent/security/ClientCertAuthenticator.java]: I will take over the fix from the other PR here.

File: agent/jvm/src/main/java/org/jolokia/jvmagent/security/AllAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+import java.util.ArrayList;
+
+/**
+ * Authenticator that succeeds if all of the Authenticators it delegates to succeed.
+ *
+ * @author roland
+ * @since 26.05.14
+ */
+public class AllAuthenticator extends Authenticator {
```
Comment (rhuss) [inline agent/jvm/src/main/java/org/jolokia/jvmagent/security/AllAuthenticator.java]: Just wondering whether an AllAuthenticator makes much sense. Also, it is not used currently, so maybe for KISS reasons we simply omit it now and can easily reintroduce it we allow for a more flexible configuration where an AllAuthenticator could be selected.

**Comment by rhuss** [comment]
LGTM, thanks ;)

I will update the documentation accordingly. A new Jolokia release (1.3.4) is expected at the end of this or beginning of next week.

**Comment by rhuss** [comment]
There will be still a problem for our use case, since the client certificate will be validated still that is signed by a CA. This happens before the authenticators are called, when setting up the [HTTPS subsystem](https://github.com/rhuss/jolokia/blob/7a8249cc1e36680aee26a8ddd570520eef3e190b/agent/jvm/src/main/java/org/jolokia/jvmagent/JolokiaServer.java#L415-L415). 

So I'm afraid that its not possible to have SSL with allowing both, client cert authentication and basic authentication without a prior client cert verification.

The authenticator is used only for some extra checks, like whether the presented certificate is indeed a client certificate (and not a server certificate), and whether the enclosed principal matches a configured value.

The check, whether a client cert is presented and whether it is signed by a given CA (stored in the keystore) happens before, internally in the HttpsServer which we cant influence much except for configuring it.

**Comment by chirino** [comment]
Yeah I think that limitation is ok.  In the case a client does not have cert trusted by the CA, he should not do SSL with the cert at all.

**Comment by rhuss** [comment]
But how can you provide a Pod specific client cert signed by the OpenShift CA when running from a Pod ?

**Pull request merged**